Piper robot integration

Ke-Wang1017 opened the pull request:
## What this does
This PR integrates PIPER robot (https://global.agilex.ai/products/piper) into lerobot, so that people can teleoperate, collect demos, visualize and train policy with lerobot functions for PIPER.

Piper SDK is needed: https://github.com/agilexrobotics/piper_sdk

## How it was tested
First activate the CAN bus: `bash can_activate.sh can0 1000000`
Teleoperation: 
`python lerobot/scripts/control_robot.py teleoperate --robot-path lerobot/configs/robot/piper.yaml`
Recording:
`python lerobot/scripts/control_robot.py record --robot-path lerobot/configs/robot/piper.yam --fps 30 --repo-id $USER/piper_test --single-task True --num-episodes 50 --run-compute-stats 0`
Data Visualization:
`python lerobot/scripts/visualize_dataset_html.py --repo-id $USER/piper_test`

Examples:
- Added `piper.yaml` in `configs/robot/`.
- Added `piper.py` in `common/robot_devices/robots` 
- Added `joystick_interface.py` in `common/robot_devices/robots` 

Things to be done:

- Add leader/follower arm config, currently one arm
- Fix joint naming

Ke-Wang1017 committed:
Add video processing utilities and integrate into camera control script

- Implemented `images_to_video` function to convert image sequences into video files.
- Added `save_demonstration_video` function to save demonstration images as a video.
- Integrated video saving functionality into the camera control script, allowing for automatic video creation from captured images.
- Introduced `play_video` function to enable playback of saved video files using OpenCV.
- Updated imports in relevant files to include new video processing utilities.

These enhancements improve the functionality for recording and visualizing robot demonstrations.

Ke-Wang1017 committed:
Refactor video processing utilities and enhance camera control integration

- Improved the `images_to_video` and `save_demonstration_video` functions for better video creation from image sequences.
- Enhanced the camera control script to automatically save captured images as videos.
- Added `play_video` functionality for playback of saved videos using OpenCV.
- Cleaned up imports to ensure all necessary video processing utilities are included.

These updates streamline the video recording and visualization process for robot demonstrations.

Ke-Wang1017 committed:
Enhance joystick control and integrate intervention class for robot teleoperation

- Updated `JoystickInterface` to include a comment for the BTN_EAST button functionality.
- Introduced `JoystickIntervention` class to manage joystick actions and gripper control, improving the interaction with the robot.
- Modified `PiperRobot` to utilize `JoystickIntervention` for teleoperation, enhancing the robot's control capabilities.
- Added a `Rate` class to manage execution frequency for smoother operation.
- Refactored the `send_action` method to handle actions more effectively and updated the robot's connection logic.
- Removed the obsolete `control_robot_piper.py` script to streamline the codebase.

These changes improve the joystick control system and overall robot teleoperation experience.

Ke-Wang1017 committed:
Enhance robot control and logging functionality

- Added signal handling in the Logger class for graceful shutdown of wandb logging.
- Updated control_utils.py to include additional robot type checks and modified video writer codec for better compatibility.
- Improved joystick interface with error handling during process termination and added comments for clarity.
- Introduced a LowPassFilter class in PiperRobot for smoother control input handling.
- Updated PiperRobot configuration to include leader arm definitions and adjusted camera settings.
- Disabled camera display by default in control_robot.py to prevent freezing issues.

These changes improve the robustness and usability of the robot control system, enhancing both logging and operational stability.

Ke-Wang1017 committed:
Enhance action handling in PiperRobot to support tensor inputs

- Added a check in the PiperRobot class to convert action inputs from torch.Tensor to list format, ensuring compatibility with existing processing logic.
- This change improves the flexibility of the robot's action handling, allowing for seamless integration of tensor-based inputs.

Ke-Wang1017 committed:
Enhance PiperRobot functionality and camera integration

- Added debug print statements in the Intel RealSense camera detection to log serial numbers and names of detected cameras.
- Refactored the LowPassFilter class in PiperRobot to accept a single orientation parameter instead of separate rx, ry, and rz values, improving code clarity.
- Updated the PiperRobot class to apply the LowPassFilter to the robot's state during action execution, enhancing control stability.
- Modified the PiperRobot configuration to replace the wrist camera with an Intel RealSense D415, ensuring better camera integration and settings.

These changes improve the robot's camera handling and overall control system stability.

helper2424 commented:
@Ke-Wang1017 that's very cool. You are 🚀 

I can't test it because I don't have those hands, but can check the code. 

I think that need to cleanup a little bit PR, because it has a lot of already removed commits from `main`. If you need assistance with it - I can help.

Ke-Wang1017 commented:
> @Ke-Wang1017 that's very cool. You are 🚀
> 
> I can't test it because I don't have those hands, but can check the code.
> 
> I think that need to cleanup a little bit PR, because it has a lot of already removed commits from `main`. If you need assistance with it - I can help.

Hi @helper2424 thank you so much, yes I just realized that some commits are from hil-serl port. It will be great if you are willing to help :)

Ke-Wang1017 committed:
Refactor Euler angle handling in PiperRobot

- Replaced the LowPassFilter class with a new EulerAngles class to improve signal rectification for orientation data.
- Introduced a rectify_signal function to handle signal adjustments based on previous values, enhancing the accuracy of orientation readings.
- Updated PiperRobot to utilize the new rectification method during state updates, ensuring more stable control during operation.

These changes enhance the robot's orientation handling and improve overall control stability.

Ke-Wang1017 committed:
Update signal rectification logic in PiperRobot

- Modified the `rectify_signal` function to adjust the threshold for signal rectification from 350000 to 180000, enhancing the sensitivity of signal adjustments.
- This change aims to improve the accuracy of signal handling in the PiperRobot, contributing to better overall control and stability during operation.

helper2424 commented on lerobot/common/robot_devices/robots/piper.py:
```diff
@@ -0,0 +1,309 @@
+import torch
+import numpy as np
+from lerobot.common.robot_devices.robots.joystick_interface import JoystickIntervention, ControllerType
+from piper_sdk import *
```
It seems that need to add it as dependency to the poetry. So it could be installed via `pip` or `poetry`.

Also, it should be hidden in the `extra` block to avoid having troubles with installation. Not everybody have `piper`.

Also, the import should be hidden inside the class, by the same reason.

KeWang1017 replied:
Right, do you have suggestions how to do this properly? Will be great if you can do this part

helper2424 replied:
The Fix is here - https://github.com/Ke-Wang1017/lerobot/pull/2

helper2424 reviewed (commented):
I have added some comments on the part that checked. I will add more later.

helper2424 committed:
Add piper poetry deps

helper2424 committed:
Fixup import for piper sdk

Ke-Wang1017 committed:
Merge pull request #2 from helper2424/user/helper2424/piper_adaptation_fix_imports

Piper. Fix imports

helper2424 commented on lerobot/common/robot_devices/robots/joystick_interface.py:
```diff
@@ -0,0 +1,202 @@
+        self.controller_config = self.CONTROLLER_CONFIGS[controller_type]
+
+        # Manager to handle shared state between processes
+        self.manager = multiprocessing.Manager()
```
Nice approach, I haven't used it a lot in production - but looks great.

Ke-Wang1017 replied:
Yeah it is nice, never had issues using for 3 different computers, although you can use `multithreads` as well

helper2424 commented on lerobot/common/robot_devices/robots/joystick_interface.py:
```diff
@@ -0,0 +1,202 @@
+
+    def get_action(self):
+        """Returns the latest action and button state from the Joystick."""
+        action = self.latest_data["action"]
```
The `latest_data` keep just a last state, right?

If I call `get_action` one time per 10 seconds and during these 10 seconds do a lot of changes then I will receive only last state at the end of 10 seconds, right? Also, it should be noblocking, right?

Ke-Wang1017 replied:
Yes this is the intention

helper2424 commented on lerobot/common/robot_devices/robots/joystick_interface.py:
```diff
@@ -0,0 +1,202 @@
+        try:
+            if hasattr(self, '_process') and self._process is not None:
+                import signal
+                self._process.terminate()
```
I see that it's initialized as `self.process`, and not as `self._process`. It looks like a mistake

Ke-Wang1017 replied:
You are right, I have fixed that in a new commit

helper2424 commented on lerobot/common/robot_devices/robots/joystick_interface.py:
```diff
@@ -0,0 +1,202 @@
+    resolution: dict
+    scale: dict
+
+class JoystickInterface:
```
I have tried to check these class with my PS5

```
>>> from lerobot.common.robot_devices.robots.joystick_interface import JoystickInterface
>>> intr = JoystickInterface()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/helper2424/Documents/lerobot/lerobot/common/robot_devices/robots/joystick_interface.py", line 84, in __init__
    self.process.start()
  File "/Users/helper2424/.pyenv/versions/3.10.15/lib/python3.10/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/Users/helper2424/.pyenv/versions/3.10.15/lib/python3.10/multiprocessing/context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/Users/helper2424/.pyenv/versions/3.10.15/lib/python3.10/multiprocessing/context.py", line 288, in _Popen
    return Popen(process_obj)
  File "/Users/helper2424/.pyenv/versions/3.10.15/lib/python3.10/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/Users/helper2424/.pyenv/versions/3.10.15/lib/python3.10/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/Users/helper2424/.pyenv/versions/3.10.15/lib/python3.10/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/Users/helper2424/.pyenv/versions/3.10.15/lib/python3.10/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
TypeError: cannot pickle 'weakref.ReferenceType' object
>>> 
```

helper2424 commented on lerobot/common/robot_devices/robots/piper.py:
```diff
@@ -0,0 +1,310 @@
+        self.rate = Rate(200)
+        self.euler_filter = EulerAngles()
+        # init piper robot
+        self.piper = C_PiperInterface("can0")
```
Can be something else except `can0`? Maybe it should be a parameter?

Ke-Wang1017 replied:
That is a good idea, if there is only one arm then `can0`, if two arms can be different

helper2424 commented on lerobot/common/robot_devices/robots/joystick_interface.py:
```diff
@@ -0,0 +1,202 @@
+                self._process = None
+
+
+class JoystickIntervention():
```
A very minor comment. I would add tests for all new classes. They could cover not all cases, but maybe some basic successful flow.

Ke-Wang1017 replied:
That's really helpful, thanks @helper2424

helper2424 reviewed (commented):
I have added bunch of comments

Ke-Wang1017 committed:
fix the self._process

Ke-Wang1017 committed:
Merge remote-tracking branch 'origin/piper_adaptation' into piper_adaptation

Ke-Wang1017 committed:
Improve Piper robot home positioning and camera configuration

- Added a new `move_to_home_2()` method for returning to home position after demo
- Updated home position coordinates in `move_to_home()` method
- Changed top camera index from 2 to 8 in Piper configuration
- Modified `control_robot.py` to use `move_to_home_2()` when resetting Piper robot
- Simplified import of Piper SDK by using wildcard import

imstevenpmwork commented:
I think this would be a great opportunity to test this new feature: https://github.com/huggingface/lerobot/pull/2123

imstevenpmwork commented:
Thank you for this integration work and for iterating with reviewers. The PR is built on control_robot.py, robot_devices, and logger.py, all removed in the Hardware API redesign (#777, e23b41e79), so 9 of the 10 touched files no longer exist on main. The good news is that Piper support is now available through the plugin system (#2123): a community package already works, see https://github.com/AgRoboticsResearch/lerobot_robot_piper as confirmed in this thread. Closing as obsolete against the current API — contributions to that plugin or a fresh Robot-subclass PR remain very welcome.
